### PR TITLE
Pushnotification send open event

### LINKED ISSUE #1919: React Native. Push opened without tracking percentage
**Describe the bug**
Push notification campaign with incomplete metrics. Push open rate and total push opened have no tracking percentage.

**To Reproduce**
1. Set up  push notification environment for react native using amplify-cli
2. Create push campaign using channelType filter for the segment in pinpoint.
3. Write a campaign message and launch campaign immediately
4. Open push notification message with the react native app (tested with app in foreground, background and killed)
5. Wait a few minutes or more and then check  the campaign metrics. Open rate and total push opened have no tracking percentage.

**Expected behavior**
Push notification campaign with complete metrics after of receive and open notification message.

**Screenshots**
screenshot from 2018-10-16 17 03 25

**Smartphone and emulator android:**
 - Device: [Samsung Galaxy J7, Nexus 5]
 - OS: [6.1, 8.0]

**Dependencies:**
    "@aws-amplify/pushnotification": "^1.0.15",
    "aws-amplify": "^1.1.6",

### PR DESCRIPTION
**Pull request by powerful23**
*Issue #, if available:*
To send open campaign events when the app is opened by notification

fixes #1919 
*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/pushnotification/src/PushNotification.ts [hunk 1]
```diff
@@ -92,12 +110,34 @@ export default class PushNotification {
         this.addEventListenerForIOS(REMOTE_NOTIFICATION_RECEIVED, this.handleCampaignPush);
     }
 
+    _checkIfOpenedByCampaign(nextAppState) {
+        const currentState = AppState.currentState;
```
Comment (shinnoki) [inline packages/pushnotification/src/PushNotification.ts]: In this way `currentState` and `nextAppState` are always equal.
`PushNotificationIOS.getInitialNotification()` does not change from wake up, so I suggest to remove `_checkIfOpenedByCampaign` and move L117-L123 into `initializeIOS` .

Comment (powerful23) [reply packages/pushnotification/src/PushNotification.ts]: @shinnoki thanks. Did you test that change in your local device to see if it works or not?

Comment (powerful23) [reply packages/pushnotification/src/PushNotification.ts]: @shinnoki I updated the pr with your advice? Could you run a test on your device again? Thanks!

**Comment by schavaLogi** [comment]
This patch works perfectly fine in android but event must be registered out side of components lifecycle . is that the expected behaviour ?

**Comment by shaliko** [comment]
@powerful23 How I can install this patch was not able to fetch it from "aws-amplify/pushnotification" from subfolder. The only one way I found is create own repo with only one "aws-amplify/pushnotification" package, but it not works - just stop open app by click on push and no errors/logs.

**Comment by powerful23** [comment]
@shaliko you can try this under your forked amplify repo: 
```
git checkout -b powerful23-pushnotification-send-open-event master
git pull https://github.com/powerful23/aws-amplify.git pushnotification-send-open-event
```

**Comment by shaliko** [comment]
@powerful23 There is issue with targetSdkVersion: 26 => "BroadcastQueue: Background execution not allowed"

https://commonsware.com/blog/2017/04/11/android-o-implicit-broadcast-ban.html

If target switched to the 25 it starts work.

**Comment by powerful23** [comment]
@shaliko thanks

**Comment by schavaLogi** [comment]
@shaliko @powerful23 need help with a local copy to verify react-native based applications.

Added in package.json like below with cloned repo and branch pull. 
"aws-amplify": "file:../amplify-js/packages/aws-amplify/",
"aws-amplify-react-native": "file:../amplify-js/packages/pushnotification",

**Error message:**
aws-amplify/package.json` was successfully found. However, this package itself specifies a `main` module field that could not be resolved (`/node_modules/aws-amplify/lib/index.js`. Indeed, none of these files exist

**Comment by schavaLogi** [comment]
@shaliko @powerful23 
 i followed steps as per this [https://github.com/aws-amplify/amplify-js/wiki/Local-Development](url) as well and changed as dependencies as below. In this case android folder missing and not able to link.

"aws-amplify": "../amplify-js/packages/aws-amplify/",
 "aws-amplify-react-native": "../amplify-js/packages/aws-amplify-react-native/",

**Comment by powerful23** [comment]
@schavaLogi when you have the local copy of this pr, you need to do the following steps:
1. under the path of `amplify-js`
```
yarn && yarn bootstrap
yarn build
```
2. under your app path:
```
yarn add path_to_aws-amplify
yarn add path_to_push_notitification
```

**Comment by schavaLogi** [comment]
> @schavaLogi when you have the local copy of this pr, you need to do the following steps:
> 
> 1. under the path of `amplify-js`
> 
> ```
> yarn && yarn bootstrap
> yarn build
> ```
> 1. under your app path:
> 
> ```
> yarn add path_to_aws-amplify
> yarn add path_to_push_notitification
> ```
Yes i did exactly the same and added in package.json like below
"aws-amplify": "../amplify-js/packages/aws-amplify/",
"aws-amplify-react-native": "../amplify-js/packages/aws-amplify-react-native/",

Observation: 
Under  /node_modiles/aws-amplify-react-native path PushNotification & android folder not present but when we install published packages these folders present 

Do I miss anything specific to PushNotification or need to add pushnotification as a seperate packe and link?

**Comment by powerful23** [comment]
@schavaLogi the `pushNotification` package is no longer under `aws-amplify-react-native`, you need to install it explicitly and the the package name is `@aws-amplify/pushnotification`

**Comment by schavaLogi** [comment]
> @schavaLogi the `pushNotification` package is no longer under `aws-amplify-react-native`, you need to install it explicitly and the the package name is `@aws-amplify/pushnotification`

@powerful23 
Sorry I am not able to make it work and getting below errors
1. @aws-amplify/pushnotification/node_modules/.bin/rimraf'. Couldn't follow symbolic link
2. Duplicate module name: react-animated because of version mismatch. We are using RN 56

Troubleshoot: I tried "yarn link" and cleaned node_modules

**Comment by powerful23** [comment]
@schavaLogi I would suggest installing the package by using the local path instead of `yarn link`.
The steps are:
```
1. clean your node modules
2. install `aws-amplify` and `@aws-amplify/pushnotification`(using local path if it's from your local)
3. react-native link the `@aws-amplify/pushnotification` if you didn't
```

**Comment by schavaLogi** [comment]
> @schavaLogi I would suggest installing the package by using the local path instead of `yarn link`.
> The steps are:
> 
> ```
> 1. clean your node modules
> 2. install `aws-amplify` and `@aws-amplify/pushnotification`(using local path if it's from your local)
> 3. react-native link the `@aws-amplify/pushnotification` if you didn't
> ```
@powerful23 
I tried local path install as well. In case of ./gradlew installDebug no issue because it's not bundled jsAssests but in case of ./gradlew installRelease we face this error.

**Error Information:**
* What went wrong:
Failed to capture snapshot of input files for task ':app:bundleReleaseJsAndAssets' property '$1' during up-to-date check.
> Could not list contents of '/node_modules/@aws-amplify/pushnotification/node_modules/.bin/rimraf'. Couldn't follow symbolic link.

I would like to know is there any documentation to verify the patches locally so that i will follow those steps.
If possible can you please try locally with a simple application and guide me for release builds.

**Comment by schavaLogi** [comment]
@powerful23 I used npm pack in each individual package and installed the tar it's working
Thanks for your guidance.

**Pull request merged**